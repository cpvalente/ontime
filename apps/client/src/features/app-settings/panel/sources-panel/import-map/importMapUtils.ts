import { defaultImportMap, ImportMap } from 'ontime-utils';

export type TableEntry = { label: string; ontimeName: keyof ImportMap | string; importName: string };

function makeOntimeFields(importOptions: ImportMap): TableEntry[] {
  return [
    { label: 'Worksheet', ontimeName: 'worksheet', importName: importOptions.worksheet },
    { label: 'Start time', ontimeName: 'timeStart', importName: importOptions.timeStart },
    { label: 'End Time', ontimeName: 'timeEnd', importName: importOptions.timeEnd },
    { label: 'Duration', ontimeName: 'duration', importName: importOptions.duration },
    { label: 'Warning Time', ontimeName: 'timeWarning', importName: importOptions.timeWarning },
    { label: 'Danger Time', ontimeName: 'timeDanger', importName: importOptions.timeDanger },
    { label: 'Cue', ontimeName: 'cue', importName: importOptions.cue },
    { label: 'Colour', ontimeName: 'colour', importName: importOptions.colour },
    { label: 'Title', ontimeName: 'title', importName: importOptions.title },
    { label: 'Presenter', ontimeName: 'presenter', importName: importOptions.presenter },
    { label: 'Subtitle', ontimeName: 'subtitle', importName: importOptions.subtitle },
    { label: 'Note', ontimeName: 'note', importName: importOptions.note },
    { label: 'Is Public', ontimeName: 'isPublic', importName: importOptions.isPublic },
    { label: 'Skip', ontimeName: 'skip', importName: importOptions.skip },
    { label: 'Timer Type', ontimeName: 'timerType', importName: importOptions.timerType },
    { label: 'End Action', ontimeName: 'endAction', importName: importOptions.endAction },
  ];
}

function makeCustomFields(importOptions: ImportMap): TableEntry[] {
  const customFieldsImportMap: TableEntry[] = [];
  if (Object.hasOwn(importOptions, 'custom')) {
    for (const field in importOptions.custom) {
      const importName = importOptions.custom[field];
      customFieldsImportMap.push({ label: field, ontimeName: field, importName });
    }
  }
  return customFieldsImportMap;
}

export function makeImportPreview(importOptions: ImportMap) {
  const ontimeFields = makeOntimeFields(importOptions);
  const customFields = makeCustomFields(importOptions);
  return { ontimeFields, customFields };
}

export function makeImportMap(entries: TableEntry[]) {
  const importMap: ImportMap = { ...defaultImportMap };
  for (const entry of entries) {
    if (Array.isArray(entry)) {
      for (const customEntry of entry) {
        importMap.custom[customEntry.ontimeName] = customEntry.importName;
      }
    } else {
      // @ts-expect-error -- its ok
      importMap[entry.ontimeName] = entry.importName;
    }
  }
  return importMap;
}
