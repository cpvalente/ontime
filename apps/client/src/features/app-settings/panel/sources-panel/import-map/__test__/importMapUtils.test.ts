import { ImportMap } from 'ontime-utils';

import { type TableEntry, makeImportMap, makeImportPreview } from '../importMapUtils';

describe('makeImportPreview()', () => {
  it('generates preview objects from import map', () => {
    const testImportMap: ImportMap = {
      worksheet: 'event schedule',
      timeStart: 'time start',
      timeEnd: 'time end',
      duration: 'duration',
      cue: 'cue',
      title: 'title',
      presenter: 'presenter',
      subtitle: 'subtitle',
      isPublic: 'public',
      skip: 'skip',
      note: 'notes',
      colour: 'colour',
      endAction: 'end action',
      timerType: 'timer type',
      timeWarning: 'warning time',
      timeDanger: 'danger time',
      custom: {
        lighting: 'lx',
        sound: 'sound',
        video: 'av',
      },
    };
    const { ontimeFields, customFields } = makeImportPreview(testImportMap);

    expect(ontimeFields.length).toBe(16);
    expect(customFields.length).toBe(3);
    expect(customFields).toStrictEqual([
      { label: 'lighting', ontimeName: 'lighting', importName: 'lx' },
      { label: 'sound', ontimeName: 'sound', importName: 'sound' },
      { label: 'video', ontimeName: 'video', importName: 'av' },
    ]);
  });
});

describe('makeImportMap()', () => {
  it('generates an import map from a list of TableEntries', () => {
    const testEntries: TableEntry[] = [
      {
        label: 'Worksheet',
        ontimeName: 'worksheet',
        importName: 'event schedule',
      },
      {
        label: 'Start time',
        ontimeName: 'timeStart',
        importName: 'time start',
      },
      {
        label: 'End Time',
        ontimeName: 'timeEnd',
        importName: 'time end',
      },
      {
        label: 'Duration',
        ontimeName: 'duration',
        importName: 'duration',
      },
      {
        label: 'Warning Time',
        ontimeName: 'timeWarning',
        importName: 'warning time',
      },
      {
        label: 'Danger Time',
        ontimeName: 'timeDanger',
        importName: 'danger time',
      },
      {
        label: 'Cue',
        ontimeName: 'cue',
        importName: 'cue',
      },
      {
        label: 'Colour',
        ontimeName: 'colour',
        importName: 'colour',
      },
      {
        label: 'Title',
        ontimeName: 'title',
        importName: 'title',
      },
      {
        label: 'Presenter',
        ontimeName: 'presenter',
        importName: 'presenter',
      },
      {
        label: 'Subtitle',
        ontimeName: 'subtitle',
        importName: 'subtitle',
      },
      {
        label: 'Note',
        ontimeName: 'note',
        importName: 'notes',
      },
      {
        label: 'Is Public',
        ontimeName: 'isPublic',
        importName: 'public',
      },
      {
        label: 'Skip',
        ontimeName: 'skip',
        importName: 'skip',
      },
      {
        label: 'Timer Type',
        ontimeName: 'timerType',
        importName: 'timer type',
      },
      {
        label: 'End Action',
        ontimeName: 'endAction',
        importName: 'end action',
      },
      [
        {
          label: 'lighting',
          ontimeName: 'lighting',
          importName: 'lx',
        },
        {
          label: 'sound',
          ontimeName: 'sound',
          importName: 'sound',
        },
        {
          label: 'video',
          ontimeName: 'video',
          importName: 'av',
        },
      ],
    ];

    const importMap = makeImportMap(testEntries);
    expect(importMap.custom).toStrictEqual({
      lighting: 'lx',
      sound: 'sound',
      video: 'av',
    });
    expect(importMap).toMatchSnapshot();
  });
});
