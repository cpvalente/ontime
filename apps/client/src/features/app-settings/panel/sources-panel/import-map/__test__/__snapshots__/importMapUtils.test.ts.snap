// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`makeImportMap() > generates an import map from a list of TableEntries 1`] = `
{
  "colour": "colour",
  "cue": "cue",
  "custom": {
    "lighting": "lx",
    "sound": "sound",
    "video": "av",
  },
  "duration": "duration",
  "endAction": "end action",
  "isPublic": "public",
  "note": "notes",
  "presenter": "presenter",
  "skip": "skip",
  "subtitle": "subtitle",
  "timeDanger": "danger time",
  "timeEnd": "time end",
  "timeStart": "time start",
  "timeWarning": "warning time",
  "timerType": "timer type",
  "title": "title",
  "worksheet": "event schedule",
}
`;
